proposals/0000-React-Native-Test-App-as-a-Package.md

```diff
@@ -0,0 +1,285 @@
+4. Depending on the desired target platforms, the next steps are:
+   - Android:
+     - Make minimal changes to `build.gradle` (one-time cost)
+     - Open the project in Android Studio
+   - iOS:
+     - Make minimal changes to `Podfile` (one-time cost)
+     - Run `pod install`
+     - Open the newly generated Xcode workspace
+   - Windows:
+     - <@afoxman or @acoates-ms to fill in>
+   - macOS:
+     - See steps for iOS
+
+A hypothetical package should end up looking something like below:
+
+```
+my-awesome-component
+├── example
+│   ├── App.js
+│   ├── build.gradle (new/modified)
+│   ├── package.json
+│   ├── Podfile (new/modified)
+│   └── yarn.lock
+├── my-awesome-component.podspec
+├── my-awesome-component-dev.podspec (new, optional)
+├── package.json
+├── src
+└── yarn.lock
+```
+
+For a hypothetical monorepo:
+
+```
+my-monorepo
+├── package.json
+├── packages
+│   ├── my-awesome-component-android
+│   │   ├── build.gradle (new/modified)
+│   │   ├── package.json
+│   │   └── src
+│   ├── my-awesome-component-core
+│   │   ├── package.json
+│   │   └── ...
+│   └── my-awesome-component-ios
+│       ├── my-awesome-component.podspec
+│       ├── my-awesome-component-dev.podspec (new, optional)
+│       ├── package.json
+│       ├── Podfile (new/modified)
+│       └── src
+└── yarn.lock
+```
+
+Platform specific project files should no longer be checked in. They are either
+provided by React Native Test App package or are generated at build time.
+
+Once the test app package is configured, **adding an additional platform**
+should only require doing one of the platform specific substeps of step 4.
+
+**Upgrading React Native** to the latest version, or any desired version, should
+only require bumping the version of this package.
+
+This package should not require additional tools other than the ones required by
+React Native itself, and should not fundamentally change how a package is
+structured.
+
+### The Manifest
+
+The manifest will be the file where we declare all our entry points, and
+resources that should be included.
+
+We haven't decided what this manifest looks like. Our initial thought is to use
+React Native CLI's
+[configuration format](https://github.com/react-native-community/cli/blob/master/docs/configuration.md)
+as a starting point. This is what
+[auto-linking](https://github.com/react-native-community/cli/blob/master/docs/autolinking.md#how-does-it-work)
+uses today, and we can extend it to support our scenarios.
+
+For iOS (and macOS), we will need to extend the format to include additional
+native source files and assets. An Xcode project generated today, will include a
+main app target, and two test targets: One for "normal" tests, and one for UI
+tests. We will need to be able to generate `.podspec` files that can support
+all.
+
+Entry points will also be declared in the manifest. Entry points tell the test
+app how to launch your app. For instance, given the declaration below:
```

Comment:
I am wondering if the concept of `entry points` is necessary here. Wouldn't the end result be the same if `AppRegistry.registerComponent` was used directly in the source code and then, the loader just [accessed the registry](https://github.com/facebook/react-native/blob/master/Libraries/ReactNative/AppRegistry.js#L165) and read all the runnables?

I would stick to a single JavaScript file as an entry point where all the components are registered.

Reply:
`entryPoint` is more for packages that provide "screens" that are directly accessible from native (e.g. in a brownfield/hybrid app).

What you're suggesting makes a lot of sense for packages that provide components. Because they aren't apps, it's likely that the JS test app is the only registered component. There are no reasons we couldn't have both. We can read from the registry if no entry points were declared in the manifest. I'll make sure to add this in the next iteration.